packages/store/src/database.ts

```diff
@@ -68,54 +66,97 @@ export class Database {
+    const eventsByEntityID = new Map<EntityID, Event[]>();
+    for (const event of events) {
+      eventsByEntityID.set(event.entityID, [
+        ...(eventsByEntityID.get(event.entityID) ?? []),
+        event,
+      ]);
+    }
+    const eventsByEntityIDEntries = [...eventsByEntityID.entries()];
+
     const batchSize = 100;
-    for (let i = 0; i < entityIDs.length; i += batchSize) {
-      const currentEntityRecordSnapshots = await this._backend.getEntities(
-        entityIDs,
-      );
-      const newEntityRecordSnapshots = await Promise.all(
-        entityIDs
-          .slice(i, i + batchSize)
-          .map(
-            async (id) =>
-              await this._computeUpdatedEntitySnapshot(
+    for (let i = 0; i < eventsByEntityIDEntries.length; i += batchSize) {
+      const batchEntries = eventsByEntityIDEntries.slice(i, i + batchSize);
+      await this._backend.modifyEntities(
+        batchEntries.map(([id]) => id),
+        async (currentEntityRecordMap) => {
+          const newSnapshotPromises: Promise<
+            DatabaseBackendEntityRecord<Entity>
+          >[] = [];
+          for (const [id, events] of batchEntries) {
+            newSnapshotPromises.push(
+              this._computeUpdatedEntitySnapshot(
                 id,
-                currentEntityRecordSnapshots.get(id) ?? null,
+                currentEntityRecordMap.get(id) ?? null,
+                events,
               ),
-          ),
+            );
+          }
+
+          const newSnapshots = await Promise.all(newSnapshotPromises);
+          return new Map(
+            newSnapshots.map((snapshot) => [snapshot.entity.id, snapshot]),
+          );
+        },
       );
-      await this._backend.putEntities(newEntityRecordSnapshots);
     }
   }
 
   private async _computeUpdatedEntitySnapshot(
     entityID: EntityID,
-    currentRecordSnapshot: DatabaseBackendEntityRecord<Entity> | null,
+    currentRecord: DatabaseBackendEntityRecord<Entity> | null,
+    newEvents: Event[],
   ): Promise<DatabaseBackendEntityRecord<Entity>> {
-    // TODO: attempt to fast-forward by applying events incrementally on the snapshot if possible (i.e. if all the new events' client timestamps are greater than the greatest previous client timestamp)
-    const events = await this._backend.listEvents({
-      predicate: ["entityID", "=", entityID],
-    });
-    const lastEventID = events[events.length - 1];
-    if (events.length > 0) {
+    newEvents.sort((a, b) => a.timestampMillis - b.timestampMillis);
+
+    // Most of the time, new events will be ordered after old events (in terms of local client time), which means we can apply them directly on top of our existing snapshot.
+    if (
+      currentRecord &&
+      newEvents.length > 0 &&
```

Comment:
Should this be validated at the HTTP API level?

Reply:
Practically speaking, `newEvents` can never be empty because we got here by iterating a map which was built by collating events by entity.

But what was in my mind was that a method which recomputes the snapshot for an entity is generally useful, and that we might want to recompute an entity's snapshot in some other circumstance, outside `putEvents`, in which case `newEvents` would be empty, and we'd want to fetch all the events. Arguably, YAGNI… I suppose I was mostly just being defensive.